zenby/shop-api-gateway
# Task 9: Backend for Frontend

### Task
https://github.com/EPAM-JS-Competency-center/cloud-development-course-initial/blob/main/9_backend_for_frontend/task.md

### Required info

| Option  | Link |
| ------------- | ------------- |
| Frontend PR  | https://github.com/zenby/shop-react-redux-cloudfront/pull/7  |
| Frontend deployment | https://d2qv2dd00nbm2g.cloudfront.net/ |
| Frontend repo  | https://github.com/zenby/shop-react-redux-cloudfront  |
| Cart service repo  | https://github.com/zenby/rs-cart-api  |
| Products, import, bff-service repo |  https://github.com/zenby/shop-api-gateway  |
| Product Service API URL  | https://omdfergjvk.execute-api.eu-central-1.amazonaws.com/dev/products  |
| Cart Service API URL  | http://zenby-cart-api-develop.eu-central-1.elasticbeanstalk.com/api/profile/cart |
| BFF Service API URL  | https://vy0lfpdap9.execute-api.eu-central-1.amazonaws.com/ |

#### Example of how to call createProduct labmda with all needed information

POST https://vy0lfpdap9.execute-api.eu-central-1.amazonaws.com/products
Product format:
```json
{
    "amount": 11,
    "description": "Description",
    "price": 56,
    "image": "https://images.unsplash.com/photo-1615327495408-4c38bd351323?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwxfDB8MXxyYW5kb218MHx8cmVsaWdpb24sY3Jvc3N8fHx8fHwxNjYwNDEyMzQx&ixlib=rb-1.2.1&q=80&utm_campaign=api-credit&utm_medium=referral&utm_source=unsplash_source&w=1080",
    "title": "Title of a new product"
}
```
Take into account that all the titles of the products must be unique, so the best idea is to generate your own title to avoid bad request error.

#### Example how to call Product Service and Cart Service via BFF Service URL
- Product service
GET https://vy0lfpdap9.execute-api.eu-central-1.amazonaws.com/products

- Cart service
GET https://vy0lfpdap9.execute-api.eu-central-1.amazonaws.com/cart


### Application Functionality (MUST HAVE)

By this point your application must be able to do:

1. Products representation on Home page should be based on Product Service API.
![image](https://user-images.githubusercontent.com/32563455/184531684-5a5f12de-afae-4005-bff7-9b258eac9c2e.png)

3. Products are coming from Product DB.
Check https://vy0lfpdap9.execute-api.eu-central-1.amazonaws.com/products

4. Product images are not randomly generated on client side. Product image, same as another product model information should be stored on BE side in Product DB.

![image](https://user-images.githubusercontent.com/32563455/184531719-33bf4a88-4ebf-4d15-b4e8-0b737a143f34.png)

5. Products might be created through CSV product file import from client side.
Auth token
```js
localStorage.setItem('authorization_token', 'emVuYnk6VEVTVF9QQVNTV09SRA==')
```
Csv file example:
```csv
title,description,price,amount,image
TESTING ITEM1,description1,21,4,https://images.unsplash.com/photo-1508888965200-3af1b13dbf3b
TESTING ITEM2,description2,22,4,https://images.unsplash.com/photo-1528825539566-2bcb5882445c
```

![auth-import](https://user-images.githubusercontent.com/32563455/184532219-da843385-1560-4cdd-8557-5069b50df66d.gif)

6. Cart might be created with appropriate product set.
Need to say that this gif was recorded on local environment. Unfortunately I wasn't able to fix mix content issue on deployed application. But anyway except this issue functionality is working properly.

![cart](https://user-images.githubusercontent.com/32563455/184532034-672979aa-d314-4666-9471-172ce9032bae.gif)

7. Auth logic should be in place

[E2E tests](https://github.com/zenby/shop-react-redux-cloudfront/blob/main/cypress/e2e/spec.cy.ts) 


### Additional (optional) tasks
1. Add a cache at the BFF Service level for a request to the getProductsList lambda function of the Product Service. The cache should expire in 2 minutes.

>   How to test:
>   Get products list
>   Create new product
>   Get products list - result shouldn’t have new product
>   Wait more than 2 minutes
>   Get products list - result should have new product

  
  [Code prove](https://github.com/zenby/shop-api-gateway/pull/7/files#diff-757626a634afe5e30a43099fceb6c3a9de108c678f9e9fc029492d5b39e79d37R17-R26)
  [Tests](https://github.com/zenby/shop-api-gateway/pull/7/files#diff-ccbaeb8571b16ae9b74aec0be684bde6167732a3e8d3bbdc49634763e0bf0ab7R47-R72)
  
![cache2](https://user-images.githubusercontent.com/32563455/184532540-9284f7bc-1324-4e6d-b033-1b34d5d6980e.gif)

  
2. Use NestJS to create BFF Service instead of Express

--- 


Self-estimation: 7 / 7
